File: lamb-upeu-sis/src/main/java/unmsm/edu/pe/tutorias/domain/services/TutoriaService.java
package unmsm.edu.pe.tutorias.domain.services;

import unmsm.edu.pe.personas.domain.entities.Docente;
import unmsm.edu.pe.personas.domain.entities.Estudiante;
import unmsm.edu.pe.shared.response.PageResponse;
import unmsm.edu.pe.tutorias.application.dto.AsignarEnBloqueRequest;
import unmsm.edu.pe.tutorias.application.dto.AsignarEnBloqueResponse;
import unmsm.edu.pe.tutorias.application.dto.EstudianteAsignableItem;
import unmsm.edu.pe.tutorias.application.dto.TutorComboItem;
import unmsm.edu.pe.tutorias.application.dto.TutorVigenteItem;
import unmsm.edu.pe.tutorias.application.dto.TutoriaHistorialItem;

import java.util.List;
import java.util.UUID;

public interface TutoriaService {

    PageResponse<TutorComboItem> buscarTutores(String buscar, int page, int size);

    PageResponse<EstudianteAsignableItem> estudiantesAsignables(UUID facultadId, UUID programaId, Boolean conTutor, String buscar, int page, int size);

    /** Tutor vigente del estudiante (o null). */
    TutorVigenteItem tutorVigente(UUID estudianteId);

    List<TutoriaHistorialItem> historial(UUID estudianteId);

    /**
     * Núcleo reutilizable: cierra la tutoría vigente (si es de otro docente) y abre
     * una nueva; idempotente si ya es el mismo tutor; valida el cupo del docente.
     * Debe invocarse dentro de una transacción (p. ej. desde el guardado de estudiante).
     */
    void asignar(Estudiante estudiante, Docente docente, String motivoCambio);

    /** Asignación individual (resuelve entidades y aplica {@link #asignar}). */
    void asignarIndividual(UUID estudianteId, UUID tutorId, String motivoCambio);

    /** Asignación en bloque: valida cupo total y aplica cierre+apertura por estudiante. */
    AsignarEnBloqueResponse asignarEnBloque(AsignarEnBloqueRequest request);

    /**
     * Finaliza la tutoría vigente del estudiante con ese tutor (queda sin tutor,
     * conservando el historial). Valida que la vigente corresponda al tutor.
     */
    void finalizar(UUID docenteId, UUID estudianteId, String motivo);
}

File: themeforest-Fj3VzBCp-fuse-angularjs-material-design-admin-template/fuse/src/app/views/dashboard/tutorias/services/tutoria.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { inject, Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from '@/environments/environment';
import { END_POINTS } from '@/app/providers/utils/end-points';

@Injectable({ providedIn: 'root' })
export class TutoriaService {
  private _http = inject(HttpClient);
  private readonly base = environment.url + END_POINTS.tutorias.base;

  /** Autocomplete de tutores (docentes PROF_TUTOR) con cupo. */
  buscarTutores$(buscar?: string, page = 0, size = 10): Observable<any> {
    let params = new HttpParams().set('page', String(page)).set('size', String(size));
    if (buscar) params = params.set('buscar', buscar);
    return this._http.get<any>(`${this.base}/tutores`, { params });
  }

  /** Estudiantes para la asignación en bloque (programa + estado tutor + texto). */
  estudiantes$(facultadId?: string, programaId?: string, conTutor?: string, buscar?: string, page = 0, size = 20): Observable<any> {
    let params = new HttpParams().set('page', String(page)).set('size', String(size));
    if (facultadId) params = params.set('facultadId', facultadId);
    if (programaId) params = params.set('programaId', programaId);
    if (conTutor) params = params.set('conTutor', conTutor);
    if (buscar) params = params.set('buscar', buscar);
    return this._http.get<any>(`${this.base}/estudiantes`, { params });
  }

  tutorVigente$(estudianteId: string): Observable<any> {
    return this._http.get<any>(`${this.base}/estudiante/${estudianteId}/vigente`);
  }

  historial$(estudianteId: string): Observable<any> {
    return this._http.get<any>(`${this.base}/estudiante/${estudianteId}/historial`);
  }

  asignar$(estudianteId: string, tutorId: string, motivoCambio?: string): Observable<any> {
    return this._http.post<any>(`${this.base}/asignar`, { estudianteId, tutorId, motivoCambio });
  }

  asignarEnBloque$(tutorId: string, estudianteIds: string[], motivoCambio?: string): Observable<any> {
    return this._http.post<any>(`${this.base}/asignar-en-bloque`, { tutorId, estudianteIds, motivoCambio });
  }

  /** Quitar un estudiante de un tutor (finaliza su tutoría vigente). */
  finalizar$(tutorId: string, estudianteId: string, motivo?: string): Observable<any> {
    let params = new HttpParams();
    if (motivo) params = params.set('motivo', motivo);
    return this._http.delete<any>(`${this.base}/${tutorId}/estudiantes/${estudianteId}`, { params });
  }
}
